File: longest-zigzag-path-in-a-binary-tree/longest-zigzag-path-in-a-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int max = 0;
    public int longestZigZag(TreeNode root) {
        dfs(root.left, 'r', 'l', 0);
        dfs(root.right, 'l', 'r', 0);
        return max;
    }

    private void dfs(TreeNode root, char lastDir, char dir, int length) {
        if (root == null) {
            return;
        }

        if (lastDir == dir) {
            length = 1;
        }
        else {
            length++;
        }

        if (length > max) {
            max = length;
        }
        dfs(root.left, dir, 'l', length);
        dfs(root.right, dir, 'r', length);
    }
}

File: 1474-longest-zigzag-path-in-a-binary-tree/longest-zigzag-path-in-a-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def longestZigZag(self, root: Optional[TreeNode]) -> int:
        dp = {}
        def helper(node, direction):
            answer = 0
            if not node:
                return 0
            if node in dp.keys() and direction in dp[node]:
                return dp[node][direction]
            if direction == 'left':
                answer = 1 + helper(node.right, 'right')
            else:
                answer = 1 + helper(node.left, 'left')
            if node in dp.keys():
                dp[node][direction] = answer
            else:
                dp[node] = {}
                dp[node][direction] = answer
            return answer
        def dfs(node):
            if not node:
                return
            helper(node, 'left')
            helper(node, 'right')
            dfs(node.left)
            dfs(node.right)
        dfs(root)
        result = 0
        for node in dp.keys():
            for direction in dp[node]:
                result = max(result, dp[node][direction])
        return result - 1